src/LagrangianParticleTracking/calculate_particle_tendency_contributions.jl

```diff
@@ -0,0 +1,84 @@
+function calculate_particle_tendency_contributions!(particles, model, active_properties)
+    arch = model.architecture
+
+    workgroup = min(length(particles), 256)
+    worksize = length(particles)
+    
+    Gp_kernel! = calculate_particle_tendency_kernel!(device(arch), workgroup, worksize)
+
+    events = []
+
+    for field in active_properties
+        property = getproperty(particles.properties, field.property)
+        tendency = model.timestepper.Gⁿ[field.tracer]
+
+        Gp_event = Gp_kernel!(property, tendency, particles, model.grid, dependencies = Event(device(arch)))
+
+        push!(events, Gp_event)
+    end
+
+    wait(device(arch), MultiEvent(Tuple(events)))
+
+    return nothing
+end
+
+calculate_particle_tendency_contributions!(model) = calculate_particle_tendency_contributions!(model.particles, model)
+calculate_particle_tendency_contributions!(particles::Nothing, model) = nothing
+calculate_particle_tendency_contributions!(particles, model) = !isnothing(particles.parameters) ? (:active_properties in keys(particles.parameters) ? calculate_particle_tendency_contributions!(particles, model, particles.parameters.active_properties) : nothing) : nothing
+calculate_particle_tendency_contributions!(particles, model, active_fields::Nothing) = nothing
+
+@inline function get_mask(x, y, z, grid, loc)
+    i, j, k = fractional_indices(x, y, z, loc, grid)
+
+    # Convert fractional indices to unit cell coordinates 0 <= (ξ, η, ζ) <=1
+    # and integer indices (with 0-based indexing).
+    ξ, i = modf(i)
+    η, j = modf(j)
+    ζ, k = modf(k)
+
+    if (ξ, η, ζ) == (0, 0, 0) #particle on grid point special case
+        return ((Int(i+1), Int(j+1), Int(k+1), 1), ), 1.0
+    else
+        nodes = repeat([(1, 1, 1, NaN)], 8)
+        _normfactor = 0.0
+        @unroll for n=1:8
+            # Move around cube corners getting node indices (0 or 1) and distances to them
+            # Distance is d when the index is 0, or 1-d when it is 1
+            a = 0^(1+(-1)^n)
+            di = 0^abs(1-a)+ξ*(-1)^a
+
+            b = 0^(1+(-1)^floor(n/2))
+            dj = 0^abs(1-b)+η*(-1)^b
+            
+            c = 0^(1+(-1)^floor(n/4))
+            dk = 0^abs(1-c)+ζ*(-1)^c
+
+            nodes[n] = (Int(i+1)+a, 
+                            Int(j+1)+b, 
+                            Int(k+1)+c, 
+                            sqrt(di^2+dj^2+dk^2))
+            _normfactor += 1 ./sqrt(di^2+dj^2+dk^2)
+        end
+        return nodes, 1/_normfactor
+    end
+end
+
+@inline get_node(::Bounded, i, N) = min(max(i, 1), N)
+@inline get_node(::Periodic, i, N) = i < 1 ? N - i : (i > N ? 1 + i : i)
```

Comment:
You may need to use `ifelse` here; otherwise it likely will not be performant (especially on GPU).

Reply:
Okay, will look into this